doc/api/esm.md

```diff
@@ -538,6 +538,43 @@ and in a CommonJS one. For example, this code will also work:
 const { something } = require('a-package/foo'); // Loads from ./foo.js.
 ```
 
+### Internal package imports
+
+In addition to the `"exports"` field it is possible to define internal package
+import maps that only apply to import specifiers from within the package itself.
+
+Entries in the imports field must always start with `#` to ensure they are
+clearly disambiguated from package specifiers.
+
+For example, the imports field can be used to gain the benefits of conditional
+exports for internal modules:
+
+```json
+// package.json
+{
+  "imports": {
+    "#dep": {
+      "node": "dep-node-native",
+      "default": "./dep-polyfill.js"
+    }
+  },
+  "dependencies": {
+    "dep-node-native": "^1.0.0"
+  }
+}
+```
+
+where `import '#dep'` would now get the resolution of the external package
+`dep-node-native` (including its exports in turn), and instead get the local
+file `./dep-polyfill.js` relative to the package in other environments.
+
+Unlike the exports field, import maps permit mapping to external packages
+because this provides an important use case for conditional loading and also can
+be done without the risk of cycles, unlike for exports.
+
+Apart from the above, the resolution rules for the imports field are otherwise
+analogous to the exports field.
```

Comment:
we might want to add that fragments and queries are not preserved for non-directory paths.

Reply:
I assume they are neither preserved nor removed..? It may make sense to call out that for bare specifiers `?`/`#`/etc. are treated like any other character if we don't do that already.

Reply:
looking at the markdown algorithm they would be removed it seems.

Reply:
I tried to find the part where it removes anything from the specifier and I can't find a reference to "?"/"#" or "query"/"fragment". According to my reading of the markdown algorithm (and memory of the spec), the following should work:

```js
// ./package.json
{
  "name": "pkg",
  "imports": {
    "#/some?private#specifier": "dep/this?is?all#just%chars"
  }
}

// ./entry.mjs
import '#/some?private#specifier';

// ./node_modules/dep/package.json
{
  "name": "pkg",
  "exports": {
    "./this?is?all#just%chars": "./foo.mjs?some=query"
  }
}
```

The import in `entry.mjs` should load `./node_modules/dep/package.json/foo.mjs?some=query`. It does honor the entire specifier, including the bits that happen to follow a question mark. So "not preserved" could be misunderstood I think. Similar behavior can also be observed in directory "exports", so this isn't specific to exact match:

```
{
  "exports": {
    "./this?may#contain/other/chars/": "./lib/"
  }
}
```

Reply:
the point is that

```json
{
  "field (you used exports but this PR is imports)": {
    "#/x?y" : "./x"
  }
```

```mjs
import '#/x?z';
```

would not carry `?` or `#` parts of the URL from input to right hand side. It also needs those to match the left hand side if they are present. I am not willing to state that these left hand side strings are not URL based/like as that goes back to my problem with putting `#` in the exports field of making the switch of being an opaque value/string quite confusing and not matching any other specifier. I do think this is a difference that was fine in "exports" since it is a single location/workflow for the exception but as we grow the usage of these non relative non propagating specifiers we need to document the actual workflow. Even reading your comment above I am unclear on your meaning which is concerning on its own.

Reply:
@bmeck like import maps, the left hand side is always a direct string replacement, so the only difference remains the standard CJS v ESM resolvers handling resolution differently for those cases, but this applies equally to any absolute or relative paths - imports and exports targets are no different.

@jkrems not that I have disallowed `#/` as a mapping prefix explicitly so your example wouldn't work and instead give an invalid specifier error. I thought it seemed sensible to lock down these base cases, but happy to reconsider.